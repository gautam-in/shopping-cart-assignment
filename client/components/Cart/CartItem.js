import styles from '../../styles/cartItem.module.scss';

const CartItem = ({ name, imageURL, price, id, quantity, inc, dec }) => {


    return (
        <div className={styles.cartItem}>
            <div className={styles.cartInfo}>
                <div className={styles.image}>
                    <img src={imageURL} alt={name} />
                </div>
                <div className={styles.info}>
                    <p>{name}</p>
                    <div className={styles.quantity}>
                        <div className={styles.incBtn} onClick={() => dec(id)}>-</div>
                        {quantity}
                        <div className={styles.decBtn} onClick={() => inc(id)}>+</div>
                        <p> {`x Rs.${price}`} </p>
                    </div>
                </div>
            </div>
            <div className={styles.total}>
                <p>Rs.{+quantity * +price}</p>
            </div>
        </div>
    )
}

export default CartItem
