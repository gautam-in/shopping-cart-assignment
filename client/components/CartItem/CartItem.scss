$primary-color:#d00156;

.cart-item {
    padding: 1em 1em;
    background-color: white;
    display: flex;
    justify-content: space-between;
    align-items: center;

    .item-name {
        font-weight: bold;
    }

    .purchase-description {
        flex-grow: 1;
        padding: 0 0 0 1em;

        .item-quantity {
            .quant-buttons {
                cursor: pointer;
                background: $primary-color;
                border: none;
                color: white;
                padding: 0.3em 0.8em;
                border-radius: 4px;

                &.left {
                    margin: 1em 1em 1em 0;
                }

                &.right {
                    margin: 1em;
                }
            }
        }
    }

    .item-price {
        padding: 0 1em;
        font-size: 1.2em;
    }
}