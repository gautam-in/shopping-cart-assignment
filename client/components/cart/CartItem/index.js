import React from "react";
import Image from "next/image";
import style from "../CartItem/cartitem.module.css";
import Button from "../../ui/Button";
import { useDispatchStore } from "public/store";

export default function CartItem({ item }) {
  const dispatch = useDispatchStore();
  const { price, quantity, imageURL, name, id } = item;
  const handleAction = (action, id) => {
    dispatch({ type: action, payload: { id } });
  };
  return (
    <>
      <div className={style.ItemContainer}>
        <div className={style.ImageContainer}>
          <Image src={imageURL} width={100} height={100} />
        </div>
        <div className={style.description}>
          <span>{name}</span>
          <div className={style.quantity}>
            <Button
              onClick={() => {
                handleAction("DECREAMENT", id);
              }}
            >
              -
            </Button>
            <span>{quantity}</span>
            <Button
              onClick={() => {
                handleAction("INCREAMENT", id);
              }}
            >
              +
            </Button>
            <span>x</span>
            <span>{price}</span>
          </div>
        </div>
        <div className={style.price}>{price * quantity}</div>
      </div>
    </>
  );
}
export function CartEndMessage({ message }) {
  return (
    <>
      <div className={style.ItemEmptyContainer}>
        <Image src="/static/images/lowest-price.png" width={144} height={51} />
        <span>{message}</span>
      </div>
    </>
  );
}
