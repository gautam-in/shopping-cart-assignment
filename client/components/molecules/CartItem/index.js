import Image from "next/image";
import { useCallback } from "react";
import { useCart } from "../../../global/utils/useCart";
import Button from "../../atoms/Button";
import {
  CartItemCount,
  CartItemDescription,
  CartItemWrapper,
  Container,
} from "./CartItem.styles";

const CartItem = ({ cartItem }) => {
  const { cartOpen, addCartItem, removeCartItem } = useCart();

  const handleAddCartItem = useCallback(() => {
    addCartItem(cartItem);
  });

  const handleRemoveCartItem = useCallback(() => {
    removeCartItem(cartItem.id);
  });

  return (
    <CartItemWrapper>
      <Container>
        <div>
          <Image
            src={cartItem.imageURL}
            height={50}
            width={50}
            alt={cartItem.name}
          />
        </div>
        <CartItemDescription>
          <h5>{cartItem.name}</h5>
          <CartItemCount>
            <Button
              name="remove item"
              onClick={handleRemoveCartItem}
              tabIndex={cartOpen ? "0" : "-1"}
            >
              -
            </Button>
            <span>{cartItem.count}</span>
            <Button
              name="add item"
              onClick={handleAddCartItem}
              tabIndex={cartOpen ? "0" : "-1"}
            >
              +
            </Button>
            <span>X </span> <span> Rs.{cartItem.price}</span>
          </CartItemCount>
        </CartItemDescription>
      </Container>

      <div>Rs.{+cartItem.price * +cartItem.count}</div>
    </CartItemWrapper>
  );
};

export default CartItem;
