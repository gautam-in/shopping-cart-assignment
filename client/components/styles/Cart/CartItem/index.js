import styled from "styled-components"
//CART RELATED
export const CartOverLay = styled.div`
    position: fixed;
    display: none;
    width: 100%;
    height: 100%;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    background-color: rgba(0,0,0,0.5);
    z-index: 2;
    cursor: pointer;
` 
export const CartContainer = styled.div`
    width: 33%;
    background-color: #fdfdf8;
    position: absolute;
    bottom: 0;
    right: 1.2rem;
    top: 30px;

    .noItemsInfo {
        position: absolute;
        top: 50%;
        left: 50%;
        transform: translate(-50%, -50%);
        text-align: center;

        h3 {   
            font-size: 1.2rem;
            margin-bottom: 10px;
        }
    }

    @media (max-width: 80em) {
        width: 40%;
    }


    @media (max-width: 62em) {
        width: 85%;
        left: 50%;
        transform: translate(-50%, 0);
    }
`

export const CartItemContainer = styled.div`
    display: grid;
    grid-template-areas: "image title title title"
                         "image quantity quantity price";
    margin: 1.2rem 0;
    background-color: #ddddc6;
`
export const CartBody = styled.div`
    overflow-y: scroll;
`

export const CartHeader = styled.div`
    background-color: black;
    color: #fff;
    padding: 0.8rem;
    display: flex;
    align-items: center;
    justify-content: space-between;

    .icon {
        color: #fff;
        font-size: 1.8rem;
    }
`

export const CartHeaderTitle = styled.p`
    font-size: 1.8rem;
`

export const ProductImageContainer = styled.div`
    grid-area: image;
    height: 9.6rem;
    width: 9.6rem;

    img  {
        width: 100%;
        height: 100%;
    }
`
export const ProductTitle = styled.p`
    grid-area: title;
    background-color: #ddddc6;
    font-size: 1.2rem;
    font-weight: bold;
    display: flex;
    align-items: center;
`

export const QuantitySeaction = styled.div`
    grid-area: quantity;
    display: flex;
    align-items: center;
    font-size: 1.2rem;

    button {
        font-size: 1.4rem;
        font-weight: bold;
        padding: 2px;
        cursor: pointer;
    }
`

export const ProductPrice = styled.h4`
    grid-area: price;
    display: flex;
    align-items: center;
    font-size: 1.2rem;
`

export const CheckoutBtnContainer = styled.div` 
    background-color: #e6e6e6;
    padding: 0.8rem;
    align-items: center;
    position: absolute;
    bottom: 0;
    right: 0;
    left: 0;

    p {
        text-align: center;
        font-size : 1.2rem; 
    }
`

export const CheckoutBtn = styled.button`
    border: none;
    width: 100%;
    margin-top: 1.2rem;
    padding: 0.8rem;
    border-radius: 5px;
    background-color: #e96443;
    color: #ffffff;
    font-size: 1.6rem;
    cursor: pointer;
`