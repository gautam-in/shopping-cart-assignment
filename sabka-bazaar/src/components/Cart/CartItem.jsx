import PropTypes from "prop-types";
import { Button } from "../Button/Button";

export const CartItem = ({
  name,
  imageURL,
  price,
  quantity,
  id,
  incrementQuantity,
  decrementQuantity,
}) => {
  return (
    <div className="cart-item-container">
      <div className="cart-item-img-container">
        <img
          className="cart-item-img"
          src={imageURL}
          alt={name}
          width="60"
          height="60"
        />
      </div>
      <div className="cart-item-info">
        <div className="cart-item-desc">
          <div className="cart-item-name">{name}</div>
          <div>
            <Button
              title="-"
              type="circularBtn"
              onClick={() => decrementQuantity(id)}
            />
            <span className="quantity">{quantity}</span>
            <Button
              title="+"
              type="circularBtn"
              onClick={() => incrementQuantity(id)}
            />
            <span className="quantity">x</span> Rs.{price}
          </div>
        </div>

        <div className="cart-item-total">Rs.{quantity * price}</div>
      </div>
    </div>
  );
};
CartItem.propTypes = {
  name: PropTypes.string,
  imageURL: PropTypes.string,
  price: PropTypes.number,
  quantity: PropTypes.number,
  id: PropTypes.string,
  incrementQuantity: PropTypes.func,
  decrementQuantity: PropTypes.func,
};
