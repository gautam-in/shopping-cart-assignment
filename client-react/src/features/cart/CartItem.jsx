import React from 'react'
import Col from 'react-bootstrap/esm/Col';
import Row from 'react-bootstrap/esm/Row';
import { useDispatch } from 'react-redux';
import { addToCart, removeFromCart } from './CartSlice'

import './Cart.scss';


export const CartItem = (props) => {
    const {item} = props;
    const dispatch = useDispatch()

  return (
    <div className='mb-3 cart-product'><Row>
    <Col xs={3} sm={2}>
        <img src={process.env.REACT_APP_BASE_URL + item.imageURL} width="100%" alt={item.name}/>
    </Col>
    <Col>
        <h6 className='cart-product-title'>{item.name}</h6>
        <div className='product-operator-container'>
            <button className='product-operator' onClick={() => dispatch(removeFromCart(item.id))}></button>
            <div>{item.qty}</div>
            <button className='product-operator plus' onClick={() => dispatch(addToCart(item))}></button>
            <div className='product-price'>
                <span>x </span>
                Rs. {item.price}
            </div>
        </div>
    </Col>
    <Col xs={3} className="d-flex align-items-end">
        Rs. {item.totalPrice}
    </Col>
</Row>
</div>
  )
}
