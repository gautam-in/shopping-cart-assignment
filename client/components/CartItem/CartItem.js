import React from "react";
import "./CartItem.scss";
import PropTypes from "prop-types";
import { useSelector, useDispatch } from "react-redux";
import {
  increaseQuantity,
  decreaseQuantity,
} from "../../redux/cart/cartActions";

const CartItem = ({ productId }) => {
  const dispatch = useDispatch();
  const product = useSelector(
    (state) =>
      state.products &&
      state.products.data &&
      state.products.data.find((_) => _.id === productId)
  );

  const purchaseDetails = useSelector(
    (state) =>
      state.cart &&
      state.cart.itemsAdded &&
      state.cart.itemsAdded.find((_) => _.id === productId)
  );

  const handleDecrement = () => {
    dispatch(decreaseQuantity(productId));
  };
  const handleIncrement = () => {
    dispatch(increaseQuantity(productId));
  };

  return (
    <div className="cart-item">
      <div className="item-image">
        <img src={product.imageURL} height={80} alt={product.name} />
      </div>
      <div className="purchase-description">
        <div className="item-name">{product.name}</div>
        <div className="item-quantity">
          <button
            type="button"
            aria-label="decrease quantity by one"
            className="quant-buttons left"
            onClick={handleDecrement}
          >
            -
          </button>
          <span>{purchaseDetails.quantity}</span>
          <button
            type="button"
            aria-label="increase quantity by one"
            className="quant-buttons right"
            onClick={handleIncrement}
          >
            +
          </button>{" "}
          <span aria-label="multiplied-by">x</span>
          &nbsp;Rs.{purchaseDetails.unitPrice}
        </div>
      </div>
      <div className="item-price">Rs.{purchaseDetails.totalPrice}</div>
    </div>
  );
};

export default CartItem;

CartItem.propTypes = {
  productId: PropTypes.string.isRequired,
};
