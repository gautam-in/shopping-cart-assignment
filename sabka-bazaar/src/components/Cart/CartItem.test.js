import { fireEvent, getByTestId, render } from "@testing-library/react";
import { CartItem } from "./CartItem";
describe("CartItem component", () => {
  it("should match snapshot", () => {
    const mockProps = {
      name: "",
      imageURL: "",
      price: 0,
      quantity: 0,
      id: "",
      incrementQuantity: jest.fn(),
      decrementQuantity: jest.fn(),
    };
    const container = render(<CartItem props={mockProps} />);
    expect(container).toMatchSnapshot();
  });

  // it("should increase or decrease quantity on click of buttons", () => {
  //   const { container, rerender } = render(<CartItem />);
  //   const incrementBtn = getByTestId(container, "incBtn");
  //   const quantity = getByTestId(container, "quantity");
  //   expect(quantity.textContent).toBe("0");
  //   fireEvent.click(incrementBtn);
  //   expect(quantity.textContent).toBe(1);
  // });
});
