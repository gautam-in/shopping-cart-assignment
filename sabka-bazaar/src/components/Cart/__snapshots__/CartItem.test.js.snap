// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartItem component should match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="cart-item-container"
      >
        <div
          class="cart-item-img-container"
        >
          <img
            class="cart-item-img"
            height="60"
            width="60"
          />
        </div>
        <div
          class="cart-item-info"
        >
          <div
            class="cart-item-desc"
          >
            <div
              class="cart-item-name"
            />
            <div>
              <button
                class="circular-btn"
              >
                <div>
                  -
                </div>
              </button>
              <span
                class="quantity"
              />
              <button
                class="circular-btn"
              >
                <div>
                  +
                </div>
              </button>
              <span
                class="quantity"
              >
                x
              </span>
               Rs.
            </div>
          </div>
          <div
            class="cart-item-total"
          >
            Rs.
            NaN
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="cart-item-container"
    >
      <div
        class="cart-item-img-container"
      >
        <img
          class="cart-item-img"
          height="60"
          width="60"
        />
      </div>
      <div
        class="cart-item-info"
      >
        <div
          class="cart-item-desc"
        >
          <div
            class="cart-item-name"
          />
          <div>
            <button
              class="circular-btn"
            >
              <div>
                -
              </div>
            </button>
            <span
              class="quantity"
            />
            <button
              class="circular-btn"
            >
              <div>
                +
              </div>
            </button>
            <span
              class="quantity"
            >
              x
            </span>
             Rs.
          </div>
        </div>
        <div
          class="cart-item-total"
        >
          Rs.
          NaN
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
